### PR TITLE
[2/3] [EAGLE] perf: Fuse TP vocab-parallel embedding

### PR DESCRIPTION
**Pull request by mattteochen**
## Motivation

The TP-local embedding path currently emits separate mask pointwise kernels, `F.embedding`/`indexSelectSmallIndex`, and a masked fill for every draft forward.

This is part 2 of 3 extracted from #30388.

## Changes

- Add a fused Triton vocab-parallel embedding kernel for the standard unquantized path.
- Gather in-shard rows and directly write zero rows for out-of-shard token IDs.
- Clamp invalid lanes to row zero before forming the weight pointer.
- Restrict the fast path to TP > 1, contiguous CUDA int32/int64 token IDs, supported CUDA weights, and `UnquantizedEmbeddingMethod`.
- Preserve the existing CPU, NPU, TP1, quantized/custom-method, and unsupported-layout fallbacks.
- Add focused kernel tests and a benchmark, including exact int64 minimum/maximum token IDs.

## Impact

This collapses the repeated TP embedding mask, gather, and fill sequence into one kernel.

main:
image

pr:
image

## Accuracy and Speedup

Accuracy and speedup were measured on the unified three-part stack. See the full results in #30388.

## Validation

- `pre-commit run --files ...`: passed.
- `python -m py_compile` on the changed kernel and test.
- Vocab embedding CUDA suite: 22 passed.
- The pointer-safety fix is also applied to #30388, and the three-part aggregate remains tree-identical to the unified branch.

## Series

- [1/3: Fuse topk=1 draft postprocess](https://github.com/sgl-project/sglang/pull/30947)
- [2/3: Fuse TP vocab-parallel embedding](https://github.com/sgl-project/sglang/pull/30948)
- [3/3: Remove redundant draft capture input staging](https://github.com/sgl-project/sglang/pull/30949)

---
### CI States

Latest PR Test (Base): :x: [Run #29502554577](https://github.com/sgl-project/sglang/actions/runs/29502554577)
Latest PR Test (Extra): :white_check_mark: [Run #29535891092](https://github.com/sgl-project/sglang/actions/runs/29535891092)

### CODE REVIEW

File: python/sglang/srt/environ.py [hunk 1]
```diff
@@ -966,6 +947,9 @@ class Envs:
     # get_embedding_tp_kwargs); they must stay in lock-step. Currently only
     # applies to the Deepseek-V2 family (Deepseek V3.1, Kimi K2.5) + drafts.
     SGLANG_ENABLE_EMBED_REPLICATION = EnvBool(False)
+    # Vocab parallel embedding: fused Triton mask+gather+fill kernel for the
+    # TP-sharded embedding lookup (kill-switch; applies to every TP>1 model).
+    SGLANG_OPT_USE_TRITON_VOCAB_PARALLEL_EMBEDDING = EnvBool(True)
```
Comment (kpham-sgl) [inline python/sglang/srt/environ.py]: Is there a reason one would disable this for TP > 1? If no, we can remove this env var and inline it

File: python/sglang/kernels/ops/embeddings/vocab_parallel_embedding.py
Comment (kpham-sgl) [inline python/sglang/kernels/ops/embeddings/vocab_parallel_embedding.py:1]: We have an ongoing [RFC](https://github.com/sgl-project/sglang/issues/29630) for kernel migration. I think this op should live in a new `sglang/kernels/ops/embeddings`

File: python/sglang/kernels/ops/embeddings/vocab_parallel_embedding.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    num_org_vocab_padding: int,
+    added_vocab_start_index: int,
+    added_vocab_end_index: int,
+) -> torch.Tensor:
+    assert input_.is_cuda
+    assert input_.is_contiguous()
+    assert input_.dtype in (torch.int32, torch.int64)
+    assert weight.is_cuda
+    assert weight.ndim == 2
+    assert weight.stride(1) == 1
```
Comment (kpham-sgl) [inline python/sglang/kernels/ops/embeddings/vocab_parallel_embedding.py:70]: Should this be in kernel selection code too?

**Comment by kpham-sgl** [comment]
/rerun-test test/registered/kernels/test_vocab_parallel_embedding.py

**Comment by kpham-sgl** [comment]
@mattteochen can you rebase with main?

File: python/sglang/kernels/ops/embeddings/vocab_parallel_embedding.py (hunk 2 above)
Comment (mattteochen) [reply python/sglang/kernels/ops/embeddings/vocab_parallel_embedding.py:70]: Yes, I've added them to `_use_triton_embedding` too. Otherwise we can't fall back

File: python/sglang/kernels/ops/embeddings/vocab_parallel_embedding.py
Comment (mattteochen) [reply python/sglang/kernels/ops/embeddings/vocab_parallel_embedding.py:1]: Oh, nice. I'll move it. Thank you for the reference!

**Review by kpham-sgl — APPROVED** [review]
